### ISSUE TITLE
Few-shot Image Generation via Cross-domain Correspondence

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#87]
*CVPR 2021*

## Motivation

- Diverse image synthesis with only a handful (~10) training samples

## Key Idea

- Fine tune a source-domain pretrained styleGAN2 generator
- Cross-domain Distance Consistency enforces relative distance between samples in **source** and **target**  domain.
- Patch Discriminator relaxes realism and prevents over-constraint through memorization.

## Previous Approach

- TGAN (Transferring GAN) : vanilla fine-tunes on target domain
- Batch Statistics Adaptation : only adapt scale / shift params in intermediate layers
- MineGAN : transforms source latent space to be more similar to target latent
- FreezeD : freeze except for lower layers of discriminator and finetune
- Adaptive Data Augmentation
- EWC : Elastic Weight Consolidation penalizes *large* changes to source model when finetuning

### Overview
![image](https://user-images.githubusercontent.com/46308407/127976464-6deae459-1ee8-4cd7-89f7-22050887cd86.png)

- Initialize adaptation generator with source-pretrained generator
- For each latent vector, compute similarity with other N samples to obtain discrete probability distribution through softmax.
- Penalize KL divergence between source and target domain distribution for each latent code.

### Cross-domain Distance Consistency

![image](https://user-images.githubusercontent.com/46308407/127976799-4d7d3f26-f811-4904-92ef-403a72c259a7.png)
- cosine similarity for L-th layer generator activation
![image](https://user-images.githubusercontent.com/46308407/127976812-9ee45c0d-f56f-466e-8200-dba17cc8fd93.png)
- Idea from contrastive learning

### Relaxed Realism
- No guarantee that target domain discriminator knows *what's real* since it has only seen a handful of images
- It typically memorizes those 10 images and enforce strong penalty for other images, which harms diversity
- Perform image discrimination only when a latent code came from anchor set (when one-to-one correspondence exist) and perform patch discrimination otherwise. (Do not force the same global structure, but only look at style and texture)
![image](https://user-images.githubusercontent.com/46308407/127977279-14142a4b-3d7c-4fc6-8ec1-441596f3635c.png)
- Patch discriminator is a subset of image discriminator (first *k* layers) that share weights

### Final Objective
![image](https://user-images.githubusercontent.com/46308407/127977341-b5be49d0-168a-4536-a48e-d4e44c894aaa.png)
- lambda = 1000 or 10000 in practice

### Datasets
Domain | Name
----- | -----
Source | FFHQ, LSUN-Church, LSUN-Horses, LSUN-Cars
Target | Face Caricatures, Face Sketches, FFHQ-babies, FFHQ-sunclasses, landscape drawings, haunted houses, Van Gogh, Wrecked Cars, Amadeo Modigliani

### Comparison with Baseline
![image](https://user-images.githubusercontent.com/46308407/127977752-1d3931cb-6afd-4bd1-8cd9-fafaf453845b.png)
![image](https://user-images.githubusercontent.com/46308407/127977777-75bb85bd-da2e-4ca7-b32b-dc288a19a8cb.png)
![image](https://user-images.githubusercontent.com/46308407/127977832-818dcf17-49b7-4389-8c34-1c446deb47ef.png)
![image](https://user-images.githubusercontent.com/46308407/127977852-81e7996c-c5b6-49b6-840e-d7e4ab169d17.png)

- LPIPS shows how similar are those in same cluster (Baselines mostly copy-paste a given training sample)

### Ablation
![image](https://user-images.githubusercontent.com/46308407/127977960-5b76883d-4846-4753-8dc2-671839d78c1b.png)
- Distance consistency regularization and relaxed discrimination helps overcome mode collapse.
- Some images have to be inspected globally (image discriminator) for realism.
![image](https://user-images.githubusercontent.com/46308407/127978190-cbfe5cc1-c907-424b-931b-625b945d0c8a.png)

### Analysis :  Source / Target Alignment
![image](https://user-images.githubusercontent.com/46308407/127978225-1d126905-2c1d-49ac-9b5e-fbb1b810475d.png)
![image](https://user-images.githubusercontent.com/46308407/127978240-8b073aa8-af3c-40d3-9adc-de6050480750.png)
![image](https://user-images.githubusercontent.com/46308407/127978252-94d70dbd-6fee-420a-bc6c-f4391761dc74.png)

### Visual Inspection
![image](https://user-images.githubusercontent.com/46308407/127978338-d55785b2-974a-4c9a-a84d-064bcf3fd4a1.png)
![image](https://user-images.githubusercontent.com/46308407/127978412-168d3968-a379-4e95-a55e-b1bee5ed4d95.png)

## Comments
- Charming output. They must have had lots of fun trying their models on *strange* datasets like old cars and haunted houses.
- Great intuition on *preserving the relative structure* in each domain and *relaxing realism enforcement of immature discriminator*.
- Wonders if it works well for latent vectors  out of anchor set. Like truncation trick, the visual displays might be highly concentrated to some high density regions. Still, the result is eye-catching nonetheless.
